### PR TITLE
Use a crate attribute to load plugins

### LINKED ISSUE #20769: Plugins without #[no_link] produce shared library error when running
When building a project using regex_macros, the binary runs successfully on my development system, but fails on production (where rust isn't installed).

Here's the smallest program I could think of that will cause this:
```rust
#![feature(plugin)]
extern crate regex;
#[plugin]
// works correctly if #[no_link] is used here
extern crate regex_macros;

fn main() {
}
```
An even more concise example causing the same effect:
```rust
extern crate regex_macros;
fn main() {}
```
That program compiles and runs successfully on a system with rust installed, but fails on one which isn't with the following error message:
```
$ ./test-project 
./test-project: error while loading shared libraries: libstd-4e7c5e5c.so: cannot open shared object file: No such file or directory
```

Note that this works correctly if `#[no_link]` is added before `extern crate regex_macros;`. This also works correctly when regex_macros isn't used.

I wasn't sure if this is an issue in rust or the regex library, so I've submitted the issue here. Feel free to close it and direct me to the correct place to submit this.

### LINKED ISSUE #21043: Stop using `extern crate` for plugins
Then we can skip linking the library and it won't be confusing.  Perhaps

```rust
#[plugin(...any metas...)] extern crate "my-lib" as foo;
```

would become

```rust
plugin!("my-lib" ...any TTs...);
```

There's no need for `as foo` because the library is not linked into the crate at all.  The common case is just

```rust
plugin!(regex_macros);
```

when the crate name is a valid identifier and there are no arguments.

This entails making the name `plugin!` special to the macro expander, since we need to locate plugins before we can load them.

### PR DESCRIPTION
**Pull request by kmcallister**
``` rust
#[plugin] #[no_link] extern crate bleh;
```

becomes a crate attribute

``` rust
#![plugin(bleh)]
```

The feature gate is still required.

It's almost never correct to link a plugin into the resulting library / executable, because it will bring all of libsyntax and librustc with it. However if you really want this behavior, you can get it with a separate `extern crate` item in addition to the `plugin` attribute.

Fixes #21043.
Fixes #20769.

[breaking-change]

### CODE REVIEW

**Comment by brson** [comment]
Seems reasonable to me. r? @sfackler

**Comment by huonw** [comment]
This stops it being possible to load a plugin in a submodule, right? Also, does it work with `cfg_attr` to e.g. only load a plugin for testing?

**Comment by kmcallister** [comment]
> This stops it being possible to load a plugin in a submodule, right?

Yeah, which sounds good to me.

> Also, does it work with `cfg_attr` to e.g. only load a plugin for testing?

No; `cfg_attr` is a syntax extension so it runs much too late for that.  Probably a follow-up issue.

**Comment by sfackler** [comment]
Seems pretty easy to solve the cfg_attr issue by just moving the load call into the beginning of expansion.

**Comment by kmcallister** [comment]
The `cfg_attr` PR is #22116.

**Comment by kmcallister** [comment]
Rebased. @sfackler, with the `cfg_attr` PR in the queue, do you have any remaining comments about this PR?

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -284,11 +284,7 @@ impl<'a, 'v> Visitor<'v> for PostExpansionVisitor<'a> {
         }
         match i.node {
             ast::ItemExternCrate(_) => {
```
Comment (sfackler) [inline src/libsyntax/feature_gate.rs]: Can you add a message telling people what to do if they still have a `#[plugin]`-tagged extern crate?

**Comment by sfackler** [comment]
r=me with the nit. Up to you on if you want to wait and let the cfg_attr change land first or not.

**Pull request merged**